src/System.Net.WebHeaderCollection/src/System/Net/HeaderInfoTable.cs

```diff
@@ -3,88 +3,160 @@
+        private static readonly Func<string, string[]> s_setCookieParser = value => ParseValueHelper(value, isSetCookie: true);
+        private static readonly HeaderInfo s_unknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false, s_singleParser);
+        private static readonly Hashtable s_headerHashTable;
+        private static readonly Stack<char> s_valueParser = new Stack<char>();
+
+        private static string[] ParseValueHelper(string value, bool isSetCookie)
+        {
+            // RFC 6265: (for Set-Cookie header)
+            // If the name-value-pair string lacks a %x3D ("=") character, ignore the set-cookie-string entirely.
+            if (isSetCookie && (value.IndexOf('=') < 0)) return new string[0];
+
+            List<string> tempStringCollection = new List<string>();
+
+            bool inquote = false;
+            int chIndex = 0;
+            StringBuilder singleValue = new StringBuilder();
+            int insertPos = 0;
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (value[i] == '\"')
+                {
+                    inquote = !inquote;
+                }
+                else if ((value[i] == ',') && !inquote)
+                {
+                    while (s_valueParser.Count > 0) singleValue.Insert(insertPos, s_valueParser.Pop());
+                    insertPos = singleValue.Length;
+                    if (!isSetCookie || !IsDuringExpiresAttributeParsing(singleValue.ToString()))
+                    {
+                        tempStringCollection.Add(singleValue.ToString().Trim());
+                        chIndex = 0;
+                        insertPos = 0;
+                        singleValue.Clear();
+                        continue;
+                    }
+                }
+                s_valueParser.Push(value[i]);
+                chIndex++;
+            }
+
+            // Now add the last of the header values to the stringtable.
+            if (chIndex != 0)
+            {
+                while (s_valueParser.Count > 0) singleValue.Insert(insertPos, s_valueParser.Pop());
+                tempStringCollection.Add(singleValue.ToString().Trim());
+            }
+
+            s_valueParser.Clear();
+            return tempStringCollection.ToArray();
+        }
+
+        // This method is to check if we are in the middle of parsing the Expires attribute
+        // for Set-Cookie header. It needs to check two conditions: 1. If current attribute
+        // is Expires. 2. Have we finished parsing it yet. Because the Expires attribute
+        // will contain exactly one comma, no comma means we are still parsing it.
+        private static bool IsDuringExpiresAttributeParsing(string singleValue)
+        {
+            // Current cookie doesn't contain any attributes.
+            if (singleValue.IndexOf(';') < 0) return false;
+
+            string lastElement = singleValue.Split(';').Last();
```

Comment:
Is `Expires` guaranteed to be the last attribute? Isn't it possible for the Set-Cookie attributes to be specified in any order, e.g. `Set-Cookie: id=a3fWa; Expires=Wed, 21 Oct 2015 07:28:00 GMT; Secure; HttpOnly`?

Reply:
The `lastAttribute` is actually the last attribute before a comma (line 41). For Set-Cookie header, it's guaranteed to be `Expires` attribute.

Reply:
>For Set-Cookie header, it's guaranteed to be Expires attribute.

What happens if the server (or malware) sends back a malformed 'Set-Cookie' header that doesn't follow this rule?

Reply:
In this particular case (`lastElement` != Expire), our code will treat comma as delimiter for multi-value header. (normal case)

This obeys rfc 2616

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma.

Example: Key: value1,value2,value3